File: android-beans/src/test/java/rocks/frieler/android/beans/SingleBeanDependencyTest.java
package rocks.frieler.android.beans;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import rocks.frieler.android.beans.BeanDependency.Fulfillment;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;
import static rocks.frieler.android.beans.BeanDependency.Fulfillment.FULFILLED;
import static rocks.frieler.android.beans.BeanDependency.Fulfillment.UNFULFILLED;

@RunWith(MockitoJUnitRunner.class)
public class SingleBeanDependencyTest {
    @Mock
    private BeansProvider beansProvider;
    @Mock
    private SingleBeanDependencyTest bean;

    @Test
    public void testFulfillReturnsFalseWhenTheDesiredBeanIsNotAvailable() {
        SingleBeanDependency<SingleBeanDependencyTest> beanDependency = new SingleBeanDependency<>("bean", SingleBeanDependencyTest.class);
        Fulfillment fulfillment = beanDependency.fulfill(beansProvider);

        assertThat(fulfillment, is(UNFULFILLED));
    }

    @Test
    public void testFulfillReturnsTrueWhenTheDesiredBeanByTypeIsAvailable() {
        when(beansProvider.lookUpBean(SingleBeanDependencyTest.class)).thenReturn(bean);

        SingleBeanDependency<SingleBeanDependencyTest> beanDependency = new SingleBeanDependency<>(SingleBeanDependencyTest.class);
        Fulfillment fulfillment = beanDependency.fulfill(beansProvider);

        assertThat(fulfillment, is(FULFILLED));
    }

    @Test
    public void testFulfillReturnsTrueWhenTheDesiredBeanByNameAndTypeIsAvailable() {
        when(beansProvider.lookUpBean("bean", SingleBeanDependencyTest.class)).thenReturn(bean);

        SingleBeanDependency<SingleBeanDependencyTest> beanDependency = new SingleBeanDependency<>("bean", SingleBeanDependencyTest.class);
        Fulfillment fulfillment = beanDependency.fulfill(beansProvider);

        assertThat(fulfillment, is(FULFILLED));
    }

    @Test
    public void testFulfillReturnsTrueWithoutQueryingTheBeansProviderAfterItWasFulfilledAlready() {
        when(beansProvider.lookUpBean("bean", SingleBeanDependencyTest.class)).thenReturn(bean);

        SingleBeanDependency<SingleBeanDependencyTest> beanDependency = new SingleBeanDependency<>("bean", SingleBeanDependencyTest.class);
        beanDependency.fulfill(beansProvider);
        reset(beansProvider);

        Fulfillment fulfillment = beanDependency.fulfill(beansProvider);

        verifyZeroInteractions(beansProvider);
        assertThat(fulfillment, is(FULFILLED));
    }

    @Test
    public void testGetReturnsBeanAfterTheBeanDependencyWasFulfilled() {
        when(beansProvider.lookUpBean("bean", SingleBeanDependencyTest.class)).thenReturn(bean);

        SingleBeanDependency<SingleBeanDependencyTest> beanDependency = new SingleBeanDependency<>("bean", SingleBeanDependencyTest.class);
        beanDependency.fulfill(beansProvider);
        SingleBeanDependencyTest dependency = beanDependency.get();

        assertThat(dependency, is(bean));
    }
}

File: android-beans/src/test/kotlin/rocks/frieler/android/beans/SingleBeanDependencyTest.kt
package rocks.frieler.android.beans

import assertk.assertThat
import assertk.assertions.isEqualTo
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.reset
import com.nhaarman.mockitokotlin2.verifyZeroInteractions
import com.nhaarman.mockitokotlin2.whenever
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.junit.MockitoJUnitRunner
import rocks.frieler.android.beans.BeanDependency.Fulfillment

@RunWith(MockitoJUnitRunner::class)
class SingleBeanDependencyTest {
	private val beansProvider: BeansProvider = mock()
	private val bean: SingleBeanDependencyTest = mock()

	@Test
	fun `fulfill() returns UNFULFILLED when the desired bean is not available`() {
		val beanDependency = SingleBeanDependency("bean", SingleBeanDependencyTest::class)
		val fulfillment = beanDependency.fulfill(beansProvider)

		assertThat(fulfillment).isEqualTo(Fulfillment.UNFULFILLED)
	}

	@Test
	fun `fulfill() returns FULLFILLED when the desired bean by type is available`() {
		whenever(beansProvider.lookUpBean(SingleBeanDependencyTest::class)).thenReturn(bean)

		val beanDependency = SingleBeanDependency(SingleBeanDependencyTest::class)
		val fulfillment = beanDependency.fulfill(beansProvider)

		assertThat(fulfillment).isEqualTo(Fulfillment.FULFILLED)
	}

	@Test
	fun `fulfill() returns FULFILLED when the desired bean by name and type is available`() {
		whenever(beansProvider.lookUpBean("bean", SingleBeanDependencyTest::class)).thenReturn(bean)

		val beanDependency = SingleBeanDependency("bean", SingleBeanDependencyTest::class)
		val fulfillment = beanDependency.fulfill(beansProvider)

		assertThat(fulfillment).isEqualTo(Fulfillment.FULFILLED)
	}

	@Test
	fun `fulfill() returns FULFILLED without querying the BeansProvider after it was fulfilled already`() {
		whenever(beansProvider.lookUpBean("bean", SingleBeanDependencyTest::class)).thenReturn(bean)

		val beanDependency = SingleBeanDependency("bean", SingleBeanDependencyTest::class)
		beanDependency.fulfill(beansProvider)
		reset(beansProvider)
		val fulfillment = beanDependency.fulfill(beansProvider)

		verifyZeroInteractions(beansProvider)
		assertThat(fulfillment).isEqualTo(Fulfillment.FULFILLED)
	}
}
